File: leetcode/200_number_of_islands.cc
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rowSize = grid.size();
        int colSize = grid[0].size();
        int islands = 0;
        for (int r = 0; r < rowSize; r++) {
            for (int  c = 0; c < colSize; c++) {
                if (grid[r][c] == '1') {
                    bfs(r, c, grid);
                    islands++;
                }
            }
        }
        return islands;
    }
private:
    void bfs(int r, int c, vector<vector<char>>& grid) {
        int rowSize = grid.size();
        int colSize = grid[0].size();
        vector<pair<int,int>>directions = {{1, 0}, {-1 ,0}, {0, 1}, {0, -1}};
        queue<pair<int,int>>q;
        q.push({r, c});
        grid[r][c] = '0';
        while (!q.empty()) {
            auto [currRow, currCol] = q.front();
            q.pop();
            for (auto dir : directions) {
                int newRow = currRow + dir.first;
                int newCol = currCol + dir.second;
                if (newRow >= 0 && newCol >= 0 && newRow < rowSize && newCol < colSize && grid[newRow][newCol] =='1') {
                    grid[newRow][newCol] = '0';
                    q.push({newRow, newCol});
                }
            }
        }
    }
};

File: leetcode/200_number_of_islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        numRows = len(grid)
        numCols = len(grid[0])
        output = 0

        for r in range(numRows):
            for c in range(numCols):
                if grid[r][c] == '1':
                    output += 1
                    self.bfs(grid, r, c, numRows, numCols)

        return output
        
    def bfs(self, grid, r, c, numRows, numCols):
        q = []
        directions = [[1,0],[-1,0],[0,1],[0,-1]]
        q.append([r,c])
        grid[r][c] = '0'

        while q:
            row, col = q.pop()
            for dir in directions:
                newRow = row + dir[0]
                newCol = col + dir[1]
                if newRow >= 0 and newRow < numRows and newCol >= 0 and newCol < numCols and grid[newRow][newCol] == '1':
                    q.append([newRow, newCol])
                    grid[newRow][newCol] = '0'
            
